draft/0004-user-sign-in-experience.md

```diff
@@ -0,0 +1,669 @@
+
+2. High flexibility: We have identified several limitations in the current internal interaction APIs, such as the lack of support for custom user profile fields and the increasingly bloated interaction structure within the Logto server. By introducing the new experience API, we can redefine the interaction in Logto, enhancing the flexibility and sustainability of the sign-in experience.
+
+### 7.2 Alternatives
+
+Considering no extra new features or customization options are added, instead of introducing a new set of experience APIs, we could consider providing comprehensive documentation for the existing internal interaction APIs. This approach would minimize the risk of breaking changes to the sign-in experience. However, even with detailed documentation, developers may still find it challenging to understand and effectively use these internal interaction APIs.
+
+Moreover, continuing to use our existing APIs and interaction design will likely lead to a rapidly increasing maintenance effort as we keep introducing new features and flows to our system. Eventually, we will reach a point where refactoring and migrating our existing sign-in experience logic will be unavoidable. Therefore, the best time to implement these changes is now.
+
+## 8.Future possibilities
+
+### 8.1.User profile management
+
+The verification and user profile APIs can be extended to support more advanced user profile management features, outside the sign-in experience. For example, users can update their profile data, manage their MFA settings, and view their interaction history using the experience APIs through a dedicated user profile management interface.
+
+### 8.2 Simplified experience API calls
+
+Reduce the complexity of the experience API calls. We are considering merging several API calls into a single API call to simplify the user interaction flow. For example, omit the extra submit request for the sign-in experience, consider introducing a `autoSubmit` flag in some of the interaction APIs to automatically submit the interaction once all the requirements are met.
+
+### 8.3 Provide additional custom user profiles in a sign-in experience
+
+The user profile data can be extended to support custom user profile fields in the sign-in experience. Developers can define and set required custom profile fields for the user during the interaction. This will allow developers to collect additional user data and customize the user profile based on their specific use cases.
+
+### 8.4 Experience API SDK
+
+Develop a client-side SDK that simplifies the integration of the experience APIs into the front-end application. The SDK can provide a set of pre-built components and functions that handle the interaction flow, user profile management, and MFA settings. This will help developers quickly integrate the experience APIs into their applications and build a customized sign-in experience with minimal effort.
```

Comment:
```suggestion
Develop a client-side SDK that simplifies the integration of the Experience API into the front-end application. The SDK can provide a set of pre-built components and functions that handle the interaction flow, user profile management, and MFA settings. This will help developers quickly integrate the Experience API into their applications and build a customized sign-in experience with minimal effort.
```